Node verifies the peer it connects to by checking its TLS common name

shamsasari opened the pull request:
The host verification is done by a custom `NettyConnector` which doesn't accept the `NettyConnection` if the CN doesn't match up. I had to introduce a config for the network map legal name so that the node can also do the host verification check during network map registration.

shamsasari commented:
@Clintonio How do I get the CI to install the new 0.6.3 version of the plugins? It's failing because of this.

mikehearn commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -279,67 +192,155 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
+            // TODO Give some thought to the retry settings
+            retryInterval = 5.seconds.toMillis()
+            retryIntervalMultiplier = 1.5  // Exponential backoff
+            maxRetryInterval = 30.minutes.toMillis()
```
I'd suggest a lower max retry interval, like 5 minutes, or 2. The risk without backoff is you flood the target with traffic when it goes down, but a retry interval of (say) two minutes would mean in a network with 10,000 nodes all trying to send a message to the down node, you're only receiving 83 attempts to start per second which should be manageable on good hardware.

mikehearn commented on test-utils/src/main/kotlin/net/corda/testing/node/SimpleNode.kt:
```diff
@@ -0,0 +1,59 @@
+ */
+class SimpleNode(val config: NodeConfiguration, val address: HostAndPort = freeLocalHostAndPort()) : AutoCloseable {
+
+    private val databaseWithCloseable = configureDatabase(config.dataSourceProperties)
```
This could use an explicit type hint.

mikehearn commented on test-utils/src/main/kotlin/net/corda/testing/node/SimpleNode.kt:
```diff
@@ -0,0 +1,59 @@
+    private val databaseWithCloseable = configureDatabase(config.dataSourceProperties)
+    val database: Database get() = databaseWithCloseable.second
+    val userService = RPCUserServiceImpl(config)
+    val identity = generateKeyPair()
```
ditto

mikehearn commented on test-utils/src/main/kotlin/net/corda/testing/node/SimpleNode.kt:
```diff
@@ -0,0 +1,59 @@
+    val identity = generateKeyPair()
+    val executor = ServiceAffinityExecutor(config.myLegalName, 1)
+    val broker = ArtemisMessagingServer(config, address, InMemoryNetworkMapCache(), userService)
+    val networkMapRegistrationFuture: SettableFuture<Unit> = SettableFuture.create()
```
This on the other hand can be written `val foo = SettableFuture.create<Unit>()` which is shorter but just as clear, I think.

shamsasari replied:
That's what I originally went for, but it complains about platform types.

mikehearn commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingComponent.kt:
```diff
@@ -167,14 +170,70 @@ abstract class ArtemisMessagingComponent() : SingletonSerializeAsToken() {
+                    .commonName
+                // TODO Verify on the entire principle (subject)
+                if (peerLegalName != expectedCommonName) {
+                    log.warn("Peer has wrong CN - expected $expectedCommonName but got $peerLegalName")
```
Maybe this is a severe enough problem to be an error rather than warn. Also the log message should probably state outright that this is indicative of either fatal misconfiguration by the remote peer, or an SSL MITM attack!

shamsasari commented on docs/source/corda-configuration-file.rst:
```diff
@@ -15,7 +15,8 @@ and for rarely changed properties this defaulting allows the property to be excl
 ------
 
-Corda uses the Typesafe configuration library to parse the configuration see the `typesafe config on Github <https://github.com/typesafehub/config/>`_  the format of the configuration files can be simple JSON, but for the more powerful substitution features
+Corda uses the Typesafe configuration library to parse the configuration see the `typesafe config on Github <https://github.com/typesafehub/config/>`_
```
No! I'll fix it

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingComponent.kt:
```diff
@@ -167,14 +170,70 @@ abstract class ArtemisMessagingComponent() : SingletonSerializeAsToken() {
+    }
+
+    class VerifyingNettyConnectorFactory : NettyConnectorFactory() {
+        override fun createConnector(configuration: MutableMap<String, Any>?,
```
Artemis has some incredibly defensive null-checks - e.g. `configuration` can be null which doesn't really make sense. So I've played safe made them all nullable.

shamsasari commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -376,22 +379,24 @@ abstract class AbstractNode(open val configuration: NodeConfiguration, val netwo
         // Register for updates, even if we're the one running the network map.
-        updateRegistration(networkMapServiceAddress, AddOrRemove.ADD)
-        return services.networkMapCache.addMapService(net, networkMapServiceAddress, true, null)
+        return updateRegistration(networkMapAddress, ADD).flatMap { response ->
```
The function executes on the same thread that completes the original Future. I've added docs for `map` and `flatMap` accordingly.

mikehearn replied:
Yes, I know. That's not what I meant. Knowing what flatMap does isn't helpful to the reader of this code because it's usually not apparent on what thread the future will be completed on. It could be the same thread that's calling updateRegistration! Or perhaps it's a single other thread. Or perhaps it's a pool. Regardless, the reader/writer has to know in order to handle thread safety correctly.

mikehearn commented:
What has bintray got to do with this change?

Clintonio commented:
This change requires us to publish a new version of cordformation and the process for that includes publishing to bintray which is currently being reworked. It's just a temporary blocker.

shamsasari commented:
I've addressed the comments, including improving the hacky unit test. Now the network map registration Future will throw an exception if the host verification to the network map node fails. This allows that test to be a lot nicer and enables the node to be shutdown in this scenario.
